<a href="https://colab.research.google.com/github/CopotronicRifat/CS-5783-Machine-Learning-Assignments/blob/main/ML_ASSIGNMENT_1_PROBLEM_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Take x_train.npy from local drive

from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving x_train.npy to x_train.npy
User uploaded file "x_train.npy" with length 928 bytes


In [ ]:
# print data

import numpy as np
train_x = np.load(fn)
print(train_x)

[[-1.28357168]
 [-0.38295937]
 [-0.63046284]
 [ 1.8592963 ]
 [-1.50680152]
 [ 0.90135868]
 [-1.18845065]
 [ 1.27946699]
 [-2.79559268]
 [-0.14475039]
 [-2.51176662]
 [-0.92377353]
 [-2.95911812]
 [ 2.59239475]
 [ 1.47002122]
 [ 2.81860551]
 [ 1.46100659]
 [ 2.96783921]
 [-0.35850134]
 [-0.56974188]
 [-0.09267264]
 [-0.9534883 ]
 [ 1.77679785]
 [-1.00734175]
 [ 2.03823207]
 [-1.50639414]
 [-1.31528116]
 [-0.83530091]
 [ 2.73361575]
 [-2.38188316]
 [-1.96958671]
 [-1.16497692]
 [ 0.86345325]
 [ 0.02060276]
 [-1.18539218]
 [-1.44237816]
 [-1.29066957]
 [ 1.23042543]
 [-1.53563574]
 [-0.71583362]
 [-1.62457331]
 [-0.80352827]
 [ 2.54363285]
 [ 2.6201403 ]
 [ 1.1634727 ]
 [-2.88463395]
 [-1.36531844]
 [-2.98879648]
 [ 2.54305165]
 [ 0.0402963 ]
 [ 0.06302893]
 [-1.5188007 ]
 [ 0.45865999]
 [ 1.89275698]
 [-2.56055938]
 [-2.57477292]
 [-0.55834899]
 [-1.77919552]
 [ 0.48602511]
 [ 0.93916006]
 [-2.2710908 ]
 [ 2.68266704]
 [ 0.64062148]
 [-1.67091865]
 [ 2.81082109]
 [ 0.02994047]
 [ 1.60092

In [ ]:
# Take y_train.npy from local drive

from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving y_train.npy to y_train.npy
User uploaded file "y_train.npy" with length 928 bytes


In [ ]:
# print data

train_y = np.load(fn)
print(train_y)

[  6.59815531   2.61148145   1.60055092  -4.60753134  10.53345287
   9.75009482   5.17331804   5.46722989 -17.92134858   4.5766958
   7.67949377   2.39602798 -43.18986856   9.63240521   2.15157176
  32.94409902   2.31817583  56.80208657   2.77767077   1.72208031
   5.07963387   2.61185947  -3.39813927   3.06682958  -6.18080857
  10.52582866   7.11396287   1.90284658  22.5638892   13.62164877
  17.78881297   4.85302906   9.96448943   6.20595761   5.13084373
   9.33960456   6.7119859    6.22204351  11.07438186   1.5881529
  12.73926299   1.78018097   6.25750814  11.78788886   7.16704364
 -30.55413276   7.96320148 -48.79523662   6.22033038   6.40243961
   6.62828718  10.75828615   9.83499526  -5.02666831   4.62879537
   3.6465632    1.75478244  15.43626766   9.96889894   9.49535574
  16.5818648   17.31792196  10.410844    13.58707595  31.90334435
   6.29918996  -0.29958123  16.90451502   3.10323772   2.18790273
  42.15156559  -3.93438083  -9.26437274  10.40188573  -5.76900628
  -5.9875163

In [ ]:
# number of training examples

m = len(train_y)
print(m)

100


In [ ]:
# computer weight matrix

w = np.mat(np.eye(m)) 

In [ ]:
# Take x_test.npy from local drive

from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving x_test.npy to x_test.npy
User uploaded file "x_test.npy" with length 928 bytes


In [ ]:
# print data

import numpy as np
test_x = np.load(fn)
print(test_x)

[[-0.97690799]
 [ 0.82012543]
 [ 0.12518128]
 [ 0.78214297]
 [ 0.60277586]
 [-0.54793269]
 [-0.2489139 ]
 [ 0.64276662]
 [-0.42239597]
 [ 0.94844497]
 [ 0.77444081]
 [ 0.82124171]
 [-0.94766384]
 [ 0.61314217]
 [ 0.9735754 ]
 [ 0.656106  ]
 [-0.88008548]
 [-0.92394665]
 [ 0.3758878 ]
 [ 0.15459991]
 [ 0.65376051]
 [ 0.36314453]
 [-0.83470285]
 [ 0.38928717]
 [ 0.10478564]
 [-0.99690865]
 [ 0.49433077]
 [ 0.21308169]
 [ 0.23662972]
 [ 0.98813615]
 [ 0.88102549]
 [-0.30245199]
 [-0.63465604]
 [-0.70622572]
 [ 0.71387141]
 [-0.07120176]
 [ 0.58866031]
 [-0.78812441]
 [-0.43638833]
 [-0.82450625]
 [ 0.92510414]
 [-0.26902551]
 [ 0.03314189]
 [ 0.66969031]
 [ 0.29498997]
 [ 0.1120666 ]
 [ 0.37401913]
 [-0.18934502]
 [ 0.1940836 ]
 [-0.77421556]
 [-0.56905633]
 [ 0.49710487]
 [ 0.15608474]
 [ 0.58966389]
 [ 0.90569095]
 [ 0.59184605]
 [-0.25348883]
 [-0.69415043]
 [ 0.09883735]
 [ 0.47891049]
 [-0.82901489]
 [-0.1178001 ]
 [ 0.63413555]
 [ 0.84910654]
 [ 0.98248255]
 [-0.95600097]
 [ 0.22360

In [ ]:
# Take y_test.npy from local drive

from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving y_test.npy to y_test.npy
User uploaded file "y_test.npy" with length 928 bytes


In [ ]:
# print data

import numpy as np
test_y = np.load(fn)
print(test_y)

[[ 2.79966047]
 [10.15930932]
 [ 7.23615049]
 [10.28634227]
 [10.35523952]
 [ 1.78733112]
 [ 3.63328335]
 [10.41291462]
 [ 2.3654997 ]
 [ 9.42654526]
 [10.30716237]
 [10.15495991]
 [ 2.56757114]
 [10.37402552]
 [ 9.22801132]
 [10.42314394]
 [ 2.1245218 ]
 [ 2.3972125 ]
 [ 9.34150356]
 [ 7.51652654]
 [10.421677  ]
 [ 9.25464625]
 [ 1.90027325]
 [ 9.42985731]
 [ 7.03866464]
 [ 2.9723117 ]
 [10.00645696]
 [ 8.05385841]
 [ 8.26104107]
 [ 9.10480092]
 [ 9.87021131]
 [ 3.19428889]
 [ 1.59553018]
 [ 1.57993225]
 [10.4137282 ]
 [ 5.29086832]
 [10.32542245]
 [ 1.73097099]
 [ 2.28511901]
 [ 1.85796263]
 [ 9.59484604]
 [ 3.46410088]
 [ 6.33112775]
 [10.4288348 ]
 [ 8.74677543]
 [ 7.10942421]
 [ 9.32893738]
 [ 4.16061293]
 [ 7.88262478]
 [ 1.69225365]
 [ 1.72396174]
 [10.01867479]
 [ 7.53051924]
 [10.32770233]
 [ 9.72297317]
 [10.33257546]
 [ 3.59437165]
 [ 1.57311403]
 [ 6.9806587 ]
 [ 9.93557231]
 [ 1.87630698]
 [ 4.83505392]
 [10.40387081]
 [10.03491999]
 [ 9.15335071]
 [ 2.63126333]
 [ 8.14714

In [ ]:
# calculate weight matrix

tau = 1
x = 0
w = np.mat(np.eye(m)) 
weight = []
for i in range(m): 
        xi = train_x[i] 
        d = (-2 * tau * tau) 
        w[i, i] = np.exp(np.dot((xi-test_x[i]), (xi-test_x[i]).T)/d)

In [ ]:
# shapes of the matrix so far

print(train_x.shape)
print(train_y.shape)
print(w.shape)

(100, 1)
(100,)
(100, 100)


In [ ]:
# reshape to match dimention
train_y = np.reshape(train_y, (100,1))

In [ ]:
print(train_y.shape)

# calculate W*Y 
wy = np.dot(w,train_y)

print(wy.shape)

(100, 1)
(100, 1)


In [ ]:
print(train_x.shape)
print(train_y.shape)

(100, 1)
(100, 1)


In [ ]:
print(test_y)

[[ 2.79966047]
 [10.15930932]
 [ 7.23615049]
 [10.28634227]
 [10.35523952]
 [ 1.78733112]
 [ 3.63328335]
 [10.41291462]
 [ 2.3654997 ]
 [ 9.42654526]
 [10.30716237]
 [10.15495991]
 [ 2.56757114]
 [10.37402552]
 [ 9.22801132]
 [10.42314394]
 [ 2.1245218 ]
 [ 2.3972125 ]
 [ 9.34150356]
 [ 7.51652654]
 [10.421677  ]
 [ 9.25464625]
 [ 1.90027325]
 [ 9.42985731]
 [ 7.03866464]
 [ 2.9723117 ]
 [10.00645696]
 [ 8.05385841]
 [ 8.26104107]
 [ 9.10480092]
 [ 9.87021131]
 [ 3.19428889]
 [ 1.59553018]
 [ 1.57993225]
 [10.4137282 ]
 [ 5.29086832]
 [10.32542245]
 [ 1.73097099]
 [ 2.28511901]
 [ 1.85796263]
 [ 9.59484604]
 [ 3.46410088]
 [ 6.33112775]
 [10.4288348 ]
 [ 8.74677543]
 [ 7.10942421]
 [ 9.32893738]
 [ 4.16061293]
 [ 7.88262478]
 [ 1.69225365]
 [ 1.72396174]
 [10.01867479]
 [ 7.53051924]
 [10.32770233]
 [ 9.72297317]
 [10.33257546]
 [ 3.59437165]
 [ 1.57311403]
 [ 6.9806587 ]
 [ 9.93557231]
 [ 1.87630698]
 [ 4.83505392]
 [10.40387081]
 [10.03491999]
 [ 9.15335071]
 [ 2.63126333]
 [ 8.14714

In [ ]:
# add column of 1 for base parameter of theta, i.e. theta_0
train_xx = np.append(train_x, np.ones(m).reshape(m,1), axis=1)
print(train_xx.shape)

(100, 2)


In [ ]:
print(train_xx)

[[-1.28357168  1.        ]
 [-0.38295937  1.        ]
 [-0.63046284  1.        ]
 [ 1.8592963   1.        ]
 [-1.50680152  1.        ]
 [ 0.90135868  1.        ]
 [-1.18845065  1.        ]
 [ 1.27946699  1.        ]
 [-2.79559268  1.        ]
 [-0.14475039  1.        ]
 [-2.51176662  1.        ]
 [-0.92377353  1.        ]
 [-2.95911812  1.        ]
 [ 2.59239475  1.        ]
 [ 1.47002122  1.        ]
 [ 2.81860551  1.        ]
 [ 1.46100659  1.        ]
 [ 2.96783921  1.        ]
 [-0.35850134  1.        ]
 [-0.56974188  1.        ]
 [-0.09267264  1.        ]
 [-0.9534883   1.        ]
 [ 1.77679785  1.        ]
 [-1.00734175  1.        ]
 [ 2.03823207  1.        ]
 [-1.50639414  1.        ]
 [-1.31528116  1.        ]
 [-0.83530091  1.        ]
 [ 2.73361575  1.        ]
 [-2.38188316  1.        ]
 [-1.96958671  1.        ]
 [-1.16497692  1.        ]
 [ 0.86345325  1.        ]
 [ 0.02060276  1.        ]
 [-1.18539218  1.        ]
 [-1.44237816  1.        ]
 [-1.29066957  1.        ]
 

In [ ]:
print(train_xx.shape)
print(w.shape)
print(train_y.shape)

(100, 2)
(100, 100)
(100, 1)


In [ ]:
print(wy.shape)

(100, 1)


In [ ]:
# calculate X.T * W * Y

xtwy = np.dot(np.transpose(train_xx), wy)
print(xtwy.shape)

(2, 1)


In [ ]:
# calculate W * X
wx = np.dot(w, train_xx)
print(wx.shape)

(100, 2)


In [ ]:
# calculate X.T * W * X

xtwx = np.dot(np.transpose(train_xx), wx)
print(xtwx.shape)

(2, 2)


In [ ]:
# calculate inverse matrix of X.T * W * X

inv = np.linalg.pinv(xtwx)
print(inv.shape)

(2, 2)


In [ ]:
# calculate theta

theta = np.dot(inv, xtwy)
print(theta.shape)

(2, 1)


In [ ]:
print(theta)

[[-0.1816276 ]
 [ 5.31324312]]


**Problem 3.1**

We did use a basis function to calculate weight matrix. That is-

$$w_j(x) = \exp(-(x(i)- x)^2/2τ^2) $$

The basis function is Gaussian Radial Basis Function. Where **x** governs the location of the basis function. 

And, **τ** governs the spatial scale.

**Problem 3.2**

In this case of locally weighhted function, the value of the parameters will be learnt for every point based on the value of τ.

In our first problem of Question 1, we aim to predict some paramenters that can be generalized for any input to make predictions.

But in this case of question 3, paramenters are predicted locally based on the datapoints.

The bandwidth parameter, τ controls how quickly the weight should fall with the distance of the point with the training point. In simpler terms, it controls the width of how varied the data is. You can find the resemblance of the above function with the gaussian function.